firestore/src/main/java/com/firebase/ui/firestore/paging/FirestoreItemKeyedDataSource.kt

```diff
@@ -0,0 +1,235 @@
+
+    private fun startAtQuery(key: DocumentSnapshot, loadSize: Long) = directQuery.startAt(key).limit(loadSize)
+
+    private fun startAfterQuery(params: LoadParams<DocumentSnapshot>) =
+            directQuery.startAfter(params.key).limit(params.requestedLoadSize.toLong())
+
+    private fun startBeforeQuery(params: LoadParams<DocumentSnapshot>) =
+            reverseQuery.startAfter(params.key).limit(params.requestedLoadSize.toLong())
+
+    private fun endAtQuery(key: DocumentSnapshot, loadSize: Long) = reverseQuery.startAt(key).limit(loadSize)
+
+    private fun resultsListener(callback: LoadCallback<DocumentSnapshot>) = ResultsListener(callback)
+    private fun reversedResultsListener(callback: LoadCallback<DocumentSnapshot>) = ResultsListener(callback, true)
+
+    /**
+     * Implementation of [EventListener].
+     *
+     * Listen for [DocumentSnapshot] for the requested page to load.
+     * The first received result is treated as page data. The subsequent results treated as data change and invoke
+     * data source invalidation.
+     *
+     * @property callback The callback to pass data to [PagedList].
+     * @property isReversedResults The flag that tells that this listener will receive reversed result.
+     */
+    private inner class ResultsListener(
+            private val callback: LoadCallback<DocumentSnapshot>,
+            private val isReversedResults: Boolean = false
+    ) : EventListener<QuerySnapshot> {
+
+        private val isFirstEvent = AtomicBoolean(true)
+        private val taskCompletionSource = TaskCompletionSource<Unit>()
+
+        /**
+         * A [Task] that will be resolved when this listener get first result or error.
+         * Used to load initial data synchronously on stale data refresh.
+         */
+        val firstEventFiredTask get() = taskCompletionSource.task
+
+        override fun onEvent(snapshot: QuerySnapshot?, exception: FirebaseFirestoreException?) {
+            if (isFirstEvent()) {
+
+                snapshot?.documents?.also { documents ->
+                    callback.onResult(if (isReversedResults) documents.reversed() else documents)
+                } ?: notifyError(exception!!)
+
+                taskCompletionSource.setResult(Unit)
+            } else {
+                invalidate()
+            }
+        }
+
+        private inline fun isFirstEvent() = isFirstEvent.compareAndSet(true, false)
+    }
+
+    private fun notifyError(exception: FirebaseFirestoreException) {
+        getErrorCallback().onError(exception)
+    }
+
+    private inline fun getErrorCallback() = errorCallback ?: defaultErrorCallback
+
+    /**
+     * Callback to listen to data loading errors.
+     */
+    interface ErrorCallback {
+        fun onError(exception: FirebaseFirestoreException)
+    }
+
+    private inline fun <TResult> Task<TResult>.await() {
+        try {
+            Tasks.await(this)
+        } catch (e: Throwable) {  }
```

Comment:
What sorts of errors are you ignoring here and why?

Reply:
It waits for completion of the Task that tries to get key befor one that is provided on DS recreation after invalidation(that happens after any document change, creadion or delition, if this change is in the range of existing listener thar previously load page). If it fails DS just start load new portion of documents beginning from provided key.

Reply:
Sorry! It's not true right now. Described in the previous comment is how it was working at first. Now it wait for whole initial data load finished. And error handling occurred in ResultsListener.onResults method. This code only needed because if we don't wait for this load synchronously Page library erase Recyclerview by submitting empty results.